### DISCUSSION TITLE
Hash functions

### DISCUSSION THREAD

**Discussion by vanillajonathan** [PRQL/prql#1768] [Ideas]
Maybe there could be hash functions?

Like [SHA-3](https://en.wikipedia.org/wiki/SHA-3), BLAKE or ChaCha20? 

Example:
```rust
let foo = sha3_512("Hello world!")
```

**Comment by max-sixty** [comment]
This would depend on the DB. A survey of what various DBs offer for hash functions would be welcome!

**Reply by vanillajonathan** [reply]
- Microsoft SQL Server have [`HASHBYTES`](https://learn.microsoft.com/en-us/sql/t-sql/functions/hashbytes-transact-sql?view=sql-server-ver16) function which can do SHA-1 and SHA-2.
- MariaDB have [`SHA1()`](https://mariadb.com/kb/en/sha1/) and [`SHA2()`](https://mariadb.com/kb/en/sha2/).
- MySQL [encryption functions](https://dev.mysql.com/doc/refman/8.0/en/encryption-functions.html) have `sha1`, `sha2` functions.
- Oracle have the [`STANDARD_HASH`](https://docs.oracle.com/database/121/SQLRF/functions183.htm#SQLRF55647) function which can do SHA-1 and SHA-2.
- PostgreSQL [pgcrypto](https://www.postgresql.org/docs/current/pgcrypto.html) module which can do SHA-1 and SHA-2.

**Reply by max-sixty** [reply]
Ok interesting. 

There's more heterogeneity than is ideal. It's very easy to add functions to the stdlib that are common across DBs at the moment. It's a bit harder to add functions that are different; they require writing rust.

if someone would like to write something like a `sha256` function, I think we'd be quite open to that, and happy to help guide how to add it.

(Possibly I'm making it sound harder than it is — it's not difficult engineering, just less trivial than adding something like `upper`)

**Reply by vanillajonathan** [reply]
Maybe a issue should be created for this and assigned the enhancement tag?

**Reply by max-sixty** [reply]
Yes!
